mlrun/model_monitoring/applications/base.py

```diff
@@ -0,0 +1,273 @@
+        Process the monitoring event and return application results.
+
+        :param event:   (dict) The monitoring event to process.
+        :returns:       (list[ModelMonitoringApplicationResult], dict) The application results
+                        and the original event for the application.
+        """
+        resolved_event = self._resolve_event(event)
+        if not (
+            hasattr(self, "context") and isinstance(self.context, mlrun.MLClientCtx)
+        ):
+            self._lazy_init(event)
+        results = self.do_tracking(*resolved_event)
+        results = results if isinstance(results, list) else [results]
+        return results, event
+
+    def _lazy_init(self, monitoring_context: mm_context.MonitoringApplicationContext):
+        self.context = cast(mlrun.MLClientCtx, monitoring_context)
+
+    @abstractmethod
+    def do_tracking(
+        self,
+        application_name: str,
+        sample_df_stats: pd.DataFrame,
+        feature_stats: pd.DataFrame,
+        sample_df: pd.DataFrame,
+        start_infer_time: pd.Timestamp,
+        end_infer_time: pd.Timestamp,
+        latest_request: pd.Timestamp,
+        endpoint_id: str,
+        output_stream_uri: str,
+    ) -> Union[
+        mm_results.ModelMonitoringApplicationResult,
+        list[mm_results.ModelMonitoringApplicationResult],
+    ]:
+        """
+        Implement this method with your custom monitoring logic.
+
+        :param application_name:        (str) the app name
+        :param sample_df_stats:         (pd.DataFrame) The new sample distribution.
+        :param feature_stats:           (pd.DataFrame) The train sample distribution.
+        :param sample_df:               (pd.DataFrame) The new sample DataFrame.
+        :param start_infer_time:        (pd.Timestamp) Start time of the monitoring schedule.
+        :param end_infer_time:          (pd.Timestamp) End time of the monitoring schedule.
+        :param latest_request:          (pd.Timestamp) Timestamp of the latest request on this endpoint_id.
+        :param endpoint_id:             (str) ID of the monitored model endpoint
+        :param output_stream_uri:       (str) URI of the output stream for results
+
+        :returns:                       (ModelMonitoringApplicationResult) or
+                                        (list[ModelMonitoringApplicationResult]) of the application results.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def _resolve_event(
+        cls,
+        monitoring_context: mm_context.MonitoringApplicationContext,
+    ) -> tuple[
+        str,
+        pd.DataFrame,
+        pd.DataFrame,
+        pd.DataFrame,
+        pd.Timestamp,
+        pd.Timestamp,
+        pd.Timestamp,
+        str,
+        str,
+    ]:
+        """
+        Converting the event into a single tuple that will be used for passing the event arguments to the running
+        application
+
+        :param event: dictionary with all the incoming data
```

Comment:
update the docstring param
```suggestion
        :param monitoring_context: ...
```